staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/parser.go

```diff
@@ -105,6 +164,117 @@ func RewriteGeneratedGogoProtobufFile(name string, extractFn ExtractFunc, option
 	})
 }
 
+func rewriteGogoSortImport(decl ast.Decl) (string, bool) {
+	t, ok := decl.(*ast.GenDecl)
+	if !ok {
+		return "", false
+	}
+	if t.Tok != token.IMPORT {
+		return "", false
+	}
+	for _, s := range t.Specs {
+		if spec, ok := s.(*ast.ImportSpec); ok {
+			if spec.Path != nil && spec.Path.Value == `"github.com/gogo/protobuf/sortkeys"` {
+				// switch gogo sort to stdlib sort
+				spec.Path.Value = `"sort"`
+				oldName := "sortkeys"
+				if spec.Name != nil {
+					oldName = spec.Name.Name
+				}
+				spec.Name = nil
+				return oldName, true
+			}
+		}
+	}
+	return "", false
+}
+func rewriteGogoSort(decl ast.Decl, oldSortImport, newSortImport string) {
+	t, ok := decl.(*ast.FuncDecl)
+	if !ok {
+		return
+	}
+	ast.Walk(replacePackageVisitor{oldPackage: oldSortImport, newPackage: newSortImport}, t.Body)
+}
+
+var keepFuncs = map[string]bool{
+	// generated helpers
+	"sovGenerated":           true,
+	"sozGenerated":           true,
+	"skipGenerated":          true,
+	"encodeVarintGenerated":  true,
+	"valueToStringGenerated": true,
+
+	// unmarshal
+	"Reset":     true,
+	"Unmarshal": true,
+
+	// marshal
+	"Size":                 true,
+	"Marshal":              true, // TODO: drop in favor of MarshalToSizedBuffer?
+	"MarshalTo":            true, // TODO: drop in favor of MarshalToSizedBuffer?
+	"MarshalToSizedBuffer": true,
+
+	// other widely used methods
+	"String": true,
+}
+
+var keepVars = map[string]bool{
+	"ErrInvalidLengthGenerated":        true,
+	"ErrIntOverflowGenerated":          true,
+	"ErrUnexpectedEndOfGroupGenerated": true,
+}
+
+func dropUnusedGo(decl ast.Decl) bool {
+	switch t := decl.(type) {
+	case *ast.GenDecl:
+		switch t.Tok {
+		case token.IMPORT:
+			specs := []ast.Spec{}
+			for _, s := range t.Specs {
+				if spec, ok := s.(*ast.ImportSpec); ok {
+					if spec.Path == nil || !strings.HasPrefix(spec.Path.Value, `"github.com/gogo/protobuf/`) {
+						specs = append(specs, spec)
+					}
+				}
+			}
+			if len(specs) == 0 {
+				return true
+			}
```

Comment:
Is this early return correct when `len(t.Specs) > 0` and `len(specs) == 0` ?  I got lost in the logic for that part..